File: Hard/Boolean Parenthesization/boolean-parenthesization.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++
#define ll long long
const int mod = 1003;
class Solution{
public:
    int countWays(int n, string exp){
        // code here
        vector<vector<vector<ll>>> dp(n, vector<vector<ll>>(n, vector<ll>(2, 0)));
        
        for (int i = n - 1; i >= 0; i--) {
            for (int j = 0; j <= n - 1; j++) {
                // Base case 1: i > j is an invalid case, so continue.
                if (i > j) continue;
                
                for (int isTrue = 0; isTrue <= 1; isTrue++) {
                    // Base case 2: i == j, evaluate the single character.
                    if (i == j) {
                        if (isTrue == 1) dp[i][j][isTrue] = exp[i] == 'T';
                        else dp[i][j][isTrue] = exp[i] == 'F';
                        continue;
                    }
    
                    // Recurrence logic:
                    ll ways = 0;
                    for (int ind = i + 1; ind <= j - 1; ind += 2) {
                        ll lT = dp[i][ind - 1][1];
                        ll lF = dp[i][ind - 1][0];
                        ll rT = dp[ind + 1][j][1];
                        ll rF = dp[ind + 1][j][0];
    
                        if (exp[ind] == '&') {
                            if (isTrue) ways = (ways + (lT * rT) % mod) % mod;
                            else ways = (ways + (lF * rT) % mod + (lT * rF) % mod + (lF * rF) % mod) % mod;
                        }
                        else if (exp[ind] == '|') {
                            if (isTrue) ways = (ways + (lF * rT) % mod + (lT * rF) % mod + (lT * rT) % mod) % mod;
                            else ways = (ways + (lF * rF) % mod) % mod;
                        }
                        else {
                            if (isTrue) ways = (ways + (lF * rT) % mod + (lT * rF) % mod) % mod;
                            else ways = (ways + (lF * rF) % mod + (lT * rT) % mod) % mod;
                        }
                    }
                    dp[i][j][isTrue] = ways;
                }
            }
        }
        return dp[0][n - 1][1];
        }
};

//{ Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        string s;
        cin>>s;
        
        Solution ob;
        cout<<ob.countWays(n, s)<<"\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Boolean Parenthesization/boolean-parenthesization.py
#User function Template for python3
class Solution:
    def countWays(self, s):
        # code here
        n=len(s)
        def solve(s, i, j, isTrue, dp):
            if i > j:
                return 0
            if i == j:
                return 1 if s[i] == ('T' if isTrue == 1 else 'F') else 0
            if dp[i][j][isTrue] != -1:
                return dp[i][j][isTrue]
        
            temp_ans = 0
            for k in range(i + 1, j, 2):
                lt = solve(s, i, k - 1, 1, dp)
                lf = solve(s, i, k - 1, 0, dp)
                rt = solve(s, k + 1, j, 1, dp)
                rf = solve(s, k + 1, j, 0, dp)
        
                if s[k] == '&':
                    temp_ans += lt * rt if isTrue == 1 else lt * rf + lf * rt + lf * rf
                elif s[k] == '|':
                    temp_ans += lt * rt + lt * rf + lf * rt if isTrue == 1 else lf * rf
                elif s[k] == '^':
                    temp_ans += lt * rf + lf * rt if isTrue == 1 else lt * rt + lf * rf
        
            dp[i][j][isTrue] = temp_ans
            return dp[i][j][isTrue]
        dp = [[[-1 for _ in range(2)] for _ in range(n)] for _ in range(n)]
        return solve(s, 0, n - 1, 1, dp)

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        s = input().strip()
        print(Solution().countWays(s))
        print("~")

# } Driver Code Ends